homeassistant/components/rainmachine/__init__.py

```diff
@@ -456,6 +417,169 @@ async def async_unrestrict_watering(
     return True
 
 
+async def load_rainmachine_client(
+    client: Client,
+    ip_address: str,
+    entry: ConfigEntry,
+) -> None:
+    """Load the RainMachine client with the specified IP address and configuration.
+
+    This function initializes the RainMachine client with the provided IP address and
+    configuration data from the given ConfigEntry.
+
+    Args:
+        client (Client): The RainMachine client instance.
+        ip_address (str): The IP address of the RainMachine device.
+        entry (ConfigEntry): The configuration entry with RainMachine settings.
+
+    Returns:
+    None
+    """
+    await client.load_local(
+        ip_address,
+        entry.data[CONF_PASSWORD],
+        port=entry.data[CONF_PORT],
+        use_ssl=entry.data.get(CONF_SSL, DEFAULT_SSL),
+    )
+
+
+def update_controller_unique_id(
+    entry: ConfigEntry, controller: Controller, entry_updates: dict[str, Any]
+) -> None:
+    """Update the unique ID of a RainMachine controller in a configuration entry.
+
+    This function checks if the provided configuration entry already has a unique ID.
+    If it does not, it sets the unique ID to the MAC address of the RainMachine controller.
+
+    Args:
+        entry (ConfigEntry): The configuration entry to be updated.
+        controller (Controller): The RainMachine controller.
+        entry_updates (dict): A dictionary containing the updates to be applied to the entry.
+
+    Returns:
+        None
+    """
+    if not entry.unique_id or is_ip_address(entry.unique_id):
+        # If the config entry doesn't already have a unique ID, set one:
+        entry_updates["unique_id"] = controller.mac
+
+
+def move_zone_run_time_to_options(
+    entry: ConfigEntry, entry_updates: dict[str, Any]
+) -> None:
+    """Move zone run time from the config entry's data to its options.
+
+    This function checks if the configuration entry contains a zone run time in its data.
+    If a zone run time exists, it is moved from the data dictionary to the options dictionary.
+    If the `CONF_USE_APP_RUN_TIMES` option is not already set in the options, it is added and set to `False`.
+
+    Args:
+        entry (ConfigEntry): The configuration entry to be updated.
+        entry_updates (dict): A dictionary containing the updates to be applied to the entry.
+
+    Returns:
+        None
+    """
+    if CONF_DEFAULT_ZONE_RUN_TIME in entry.data:
+        # If a zone run time exists in the config entry's data, pop it and move it to options:
+        data = {**entry.data}
+        entry_updates["data"] = data
+        entry_updates["options"] = {
+            **entry.options,
+            CONF_DEFAULT_ZONE_RUN_TIME: data.pop(CONF_DEFAULT_ZONE_RUN_TIME),
+        }
+
+    if CONF_USE_APP_RUN_TIMES not in entry.options:
+        entry_updates["options"] = {**entry.options, CONF_USE_APP_RUN_TIMES: False}
+
+
+def update_config_entry(
```

Comment:
This is a one liner. You are introducing more complexity by moving it to a separate function.

Reply:
Okay, will fix that.